File: inaya-network-dapp/src/lib/s3-compat/sigv4.js
// src/lib/s3-compat/sigv4.js
//
// AWS Signature Version 4 request verification -- the one genuinely new
// cryptographic primitive in this SOW (everything else reuses existing
// Inaya crypto). Implements the real algorithm from AWS's own spec
// (canonical request -> string to sign -> derived signing key -> HMAC),
// verified against real `aws` CLI / `@aws-sdk/client-s3` requests during
// testing -- not a simplified approximation, since a real client computes
// the real thing and expects the server to match it exactly or reject it.
//
// Reference: https://docs.aws.amazon.com/general/latest/gr/sigv4-signing.html
//
// Google Cloud Storage Compatibility Layer SOW -- generalized (not forked)
// to also accept Google Cloud Storage's native GOOG4-HMAC-SHA256 scheme.
// Per the Phase 0 audit: GCS's XML API documents TWO accepted V4 signing
// modes -- "AWS4-HMAC-SHA256 with x-amz-* interoperability" (byte-identical
// to real AWS SigV4, already fully handled below with zero changes) and its
// own native "GOOG4-HMAC-SHA256" (x-goog-date/x-goog-content-sha256,
// "goog4_request" scope terminator, "GOOG4"+secret key-derivation seed --
// otherwise the exact same HMAC-chain algorithm). The two schemes are
// parameterized here as SCHEMES below rather than duplicated into a second
// file, per the SOW's own "do not fork the SigV4 engine unless a genuine
// incompatibility is proven" instruction -- there is no genuine algorithmic
// incompatibility, only different constant strings and header names.
//
// Deliberately supports header-based signing only (Authorization: AWS4-HMAC-SHA256 /
// GOOG4-HMAC-SHA256 ...), which is what the AWS CLI, every AWS SDK, gsutil/boto
// configured against a non-Google S3-compatible endpoint, and Google's own
// client libraries (when using HMAC credentials) use by default. Presigned/
// signed-URL query-string signing (`aws s3 presign`, GCS V4 signed URLs) is a
// real, separate variant this file does NOT implement -- stated here plainly
// rather than silently half-supported, per this SOW's own "no unverified
// compatibility claims" rule.

import { createHash, createHmac, timingSafeEqual } from "node:crypto";

const SERVICE = "s3";

/** The two supported signing schemes -- same algorithm, different constants.
 *  Detected from the Authorization header's own algorithm prefix, never
 *  assumed from anything else about the request. */
const SCHEMES = {
  "AWS4-HMAC-SHA256": { requestType: "aws4_request", keySeed: "AWS4", dateHeader: "x-amz-date", contentSha256Header: "x-amz-content-sha256" },
  "GOOG4-HMAC-SHA256": { requestType: "goog4_request", keySeed: "GOOG4", dateHeader: "x-goog-date", contentSha256Header: "x-goog-content-sha256" },
};

function sha256Hex(input) {
  return createHash("sha256").update(input).digest("hex");
}

function hmac(key, data) {
  return createHmac("sha256", key).update(data, "utf8").digest();
}

/** Parses the Authorization header into its Algorithm/Credential/SignedHeaders/
 *  Signature parts. Returns null (never throws) on anything malformed or an
 *  unrecognized algorithm -- callers treat that as a hard auth failure, same
 *  as a missing header. */
export function parseAuthorizationHeader(headerValue) {
  if (!headerValue) return null;
  const algorithm = Object.keys(SCHEMES).find((a) => headerValue.startsWith(a));
  if (!algorithm) return null;
  const scheme = SCHEMES[algorithm];

  const credMatch = headerValue.match(/Credential=([^,]+)/);
  const signedHeadersMatch = headerValue.match(/SignedHeaders=([^,]+)/);
  const signatureMatch = headerValue.match(/Signature=([a-f0-9]+)/);
  if (!credMatch || !signedHeadersMatch || !signatureMatch) return null;

  const credParts = credMatch[1].split("/");
  if (credParts.length !== 5 || credParts[4] !== scheme.requestType) return null;
  const [accessKeyId, date, region, service] = credParts;

  return {
    algorithm,
    scheme,
    accessKeyId,
    date,
    region,
    service,
    signedHeaders: signedHeadersMatch[1].split(";"),
    signature: signatureMatch[1],
  };
}

// URI-encodes a single path segment per RFC 3986 the way SigV4 requires --
// encodeURIComponent leaves a few characters SigV4 still wants percent-encoded.
function sigv4UriEncode(str, encodeSlash = false) {
  return encodeURIComponent(str)
    .replace(/[!'()*]/g, (c) => "%" + c.charCodeAt(0).toString(16).toUpperCase())
    .replace(/%2F/g, encodeSlash ? "%2F" : "/");
}

function canonicalUri(pathname) {
  return pathname.split("/").map((seg) => sigv4UriEncode(seg)).join("/") || "/";
}

function canonicalQueryString(searchParams) {
  const pairs = [];
  for (const [key, value] of searchParams.entries()) {
    if (key === "X-Amz-Signature") continue;
    pairs.push([sigv4UriEncode(key, true), sigv4UriEncode(value ?? "", true)]);
  }
  pairs.sort((a, b) => (a[0] < b[0] ? -1 : a[0] > b[0] ? 1 : a[1] < b[1] ? -1 : a[1] > b[1] ? 1 : 0));
  return pairs.map(([k, v]) => `${k}=${v}`).join("&");
}

function canonicalHeaders(headers, signedHeaderNames) {
  return signedHeaderNames
    .map((name) => `${name}:${(headers.get(name) || "").trim().replace(/\s+/g, " ")}\n`)
    .join("");
}

function deriveSigningKey({ secretAccessKey, date, region, service, scheme }) {
  const kDate = hmac(scheme.keySeed + secretAccessKey, date);
  const kRegion = hmac(kDate, region);
  const kService = hmac(kRegion, service);
  return hmac(kService, scheme.requestType);
}

/** Verifies a signed request -- AWS4-HMAC-SHA256 (real AWS SigV4) or
 *  GOOG4-HMAC-SHA256 (Google Cloud Storage's native V4 scheme), whichever
 *  the Authorization header itself declares (see SCHEMES above). Both are
 *  the identical HMAC-chain algorithm; only header names and a few constant
 *  strings differ. `bodyBuffer` must be the raw, unparsed request body
 *  (Buffer) -- S3/GCS PUT bodies are arbitrary binary, so this must run
 *  before any JSON/text parsing. Returns { ok: true, accessKeyId } or
 *  { ok: false, reason }. */
export function verifySigV4Request({ method, url, headers, bodyBuffer, secretAccessKey }) {
  const authHeader = headers.get("authorization");
  const parsed = parseAuthorizationHeader(authHeader);
  if (!parsed) return { ok: false, reason: "Missing or malformed Authorization header." };
  const { scheme } = parsed;

  const requestDate = headers.get(scheme.dateHeader);
  if (!requestDate) return { ok: false, reason: `Missing ${scheme.dateHeader} header.` };
  if (!requestDate.startsWith(parsed.date)) return { ok: false, reason: `${scheme.dateHeader} does not match the credential scope's date.` };

  // 15-minute clock-skew window, same tolerance every AWS SDK/GCS client itself uses.
  const requestTime = Date.parse(
    `${requestDate.slice(0, 4)}-${requestDate.slice(4, 6)}-${requestDate.slice(6, 8)}T${requestDate.slice(9, 11)}:${requestDate.slice(11, 13)}:${requestDate.slice(13, 15)}Z`
  );
  if (!Number.isFinite(requestTime) || Math.abs(Date.now() - requestTime) > 15 * 60 * 1000) {
    return { ok: false, reason: "Request timestamp is outside the allowed 15-minute window." };
  }

  const payloadHashHeader = headers.get(scheme.contentSha256Header);
  const payloadHash = payloadHashHeader === "UNSIGNED-PAYLOAD" ? "UNSIGNED-PAYLOAD" : sha256Hex(bodyBuffer || Buffer.alloc(0));
  if (payloadHashHeader && payloadHashHeader !== "UNSIGNED-PAYLOAD" && payloadHashHeader !== payloadHash) {
    return { ok: false, reason: `${scheme.contentSha256Header} does not match the actual request body.` };
  }

  const canonicalRequest = [
    method.toUpperCase(),
    canonicalUri(url.pathname),
    canonicalQueryString(url.searchParams),
    canonicalHeaders(headers, parsed.signedHeaders),
    parsed.signedHeaders.join(";"),
    payloadHash,
  ].join("\n");

  const credentialScope = `${parsed.date}/${parsed.region}/${parsed.service}/${scheme.requestType}`;
  const stringToSign = [parsed.algorithm, requestDate, credentialScope, sha256Hex(canonicalRequest)].join("\n");

  const signingKey = deriveSigningKey({ secretAccessKey, date: parsed.date, region: parsed.region, service: parsed.service || SERVICE, scheme });
  const expectedSignature = createHmac("sha256", signingKey).update(stringToSign, "utf8").digest("hex");

  const provided = Buffer.from(parsed.signature, "hex");
  const expected = Buffer.from(expectedSignature, "hex");
  const signaturesMatch = provided.length === expected.length && timingSafeEqual(provided, expected);

  if (!signaturesMatch) return { ok: false, reason: "SignatureDoesNotMatch" };
  return { ok: true, accessKeyId: parsed.accessKeyId };
}

File: inaya-drive-core/src/sigv4.rs
// src/sigv4.rs
//
// AWS Signature Version 4 request SIGNING (the client side) -- a direct
// Rust port of this same repo's src/lib/s3-compat/sigv4.js VERIFICATION
// logic, run in reverse. Both sides implement the identical real AWS
// algorithm (canonical request -> string-to-sign -> derived signing key ->
// HMAC), so a request this module signs is verified by that same,
// already-tested server code with no new server-side trust surface.

use chrono::Utc;
use hmac::{Hmac, Mac};
use sha2::{Digest, Sha256};

type HmacSha256 = Hmac<Sha256>;

fn sha256_hex(input: &[u8]) -> String {
    let mut hasher = Sha256::new();
    hasher.update(input);
    hex::encode(hasher.finalize())
}

fn hmac_bytes(key: &[u8], data: &str) -> Vec<u8> {
    let mut mac = HmacSha256::new_from_slice(key).expect("HMAC accepts any key length");
    mac.update(data.as_bytes());
    mac.finalize().into_bytes().to_vec()
}

// Percent-encodes a single path segment exactly as sigv4.js's own
// sigv4UriEncode does -- matching the server's canonicalUri() byte for
// byte is required, or the signature the server recomputes will not match.
fn sigv4_uri_encode(s: &str, encode_slash: bool) -> String {
    let mut out = String::new();
    for byte in s.bytes() {
        let c = byte as char;
        let unreserved = c.is_ascii_alphanumeric() || matches!(c, '-' | '_' | '.' | '~');
        if unreserved {
            out.push(c);
        } else if c == '/' && !encode_slash {
            out.push('/');
        } else {
            out.push_str(&format!("%{:02X}", byte));
        }
    }
    out
}

fn canonical_uri(path: &str) -> String {
    let encoded: Vec<String> = path.split('/').map(|seg| sigv4_uri_encode(seg, false)).collect();
    let joined = encoded.join("/");
    if joined.is_empty() { "/".to_string() } else { joined }
}

pub struct SignedRequest {
    pub headers: Vec<(String, String)>,
}

/// Signs one S3 REST request. `path` must already be the raw request path
/// (e.g. "/bucket/key/with/slashes"), `query_pairs` pre-sorted-by-caller is
/// NOT required -- this function sorts them itself, matching the server's
/// own canonicalQueryString(). `payload` is the raw request body.
pub fn sign(
    method: &str,
    host: &str,
    path: &str,
    query_pairs: &[(String, String)],
    access_key_id: &str,
    secret_access_key: &str,
    payload: &[u8],
) -> SignedRequest {
    let now = Utc::now();
    let amz_date = now.format("%Y%m%dT%H%M%SZ").to_string();
    let date_stamp = now.format("%Y%m%d").to_string();
    let region = "inaya";
    let service = "s3";

    let payload_hash = sha256_hex(payload);

    let mut sorted_query = query_pairs.to_vec();
    sorted_query.sort();
    let canonical_query_string = sorted_query
        .iter()
        .map(|(k, v)| format!("{}={}", sigv4_uri_encode(k, true), sigv4_uri_encode(v, true)))
        .collect::<Vec<_>>()
        .join("&");

    // Only host + x-amz-date + x-amz-content-sha256 are signed -- the
    // minimal signed-header set the server's own verifySigV4Request()
    // accepts, since it recomputes canonicalHeaders() over exactly the
    // headers named in SignedHeaders, not a fixed larger set.
    let canonical_headers = format!(
        "host:{}\nx-amz-content-sha256:{}\nx-amz-date:{}\n",
        host, payload_hash, amz_date
    );
    let signed_headers = "host;x-amz-content-sha256;x-amz-date";

    let canonical_request = format!(
        "{}\n{}\n{}\n{}\n{}\n{}",
        method.to_uppercase(),
        canonical_uri(path),
        canonical_query_string,
        canonical_headers,
        signed_headers,
        payload_hash
    );

    let credential_scope = format!("{}/{}/{}/aws4_request", date_stamp, region, service);
    let string_to_sign = format!(
        "AWS4-HMAC-SHA256\n{}\n{}\n{}",
        amz_date,
        credential_scope,
        sha256_hex(canonical_request.as_bytes())
    );

    let k_date = hmac_bytes(format!("AWS4{}", secret_access_key).as_bytes(), &date_stamp);
    let k_region = hmac_bytes(&k_date, region);
    let k_service = hmac_bytes(&k_region, service);
    let k_signing = hmac_bytes(&k_service, "aws4_request");
    let signature = hex::encode(hmac_bytes(&k_signing, &string_to_sign));

    let authorization = format!(
        "AWS4-HMAC-SHA256 Credential={}/{}, SignedHeaders={}, Signature={}",
        access_key_id, credential_scope, signed_headers, signature
    );

    // Deliberately does NOT include an explicit "Host" header here: reqwest
    // (like most HTTP clients) sets Host itself from the request URL and
    // treats it as a reserved header, so trying to also set it manually
    // risks a value that silently doesn't match what's actually put on the
    // wire. The `host` PARAMETER above is only used to compute the
    // signature's canonical_headers string -- as long as callers pass the
    // exact host:port of the URL they're about to request, the signature
    // matches what the server actually receives, with no separate header
    // needed on the wire.
    SignedRequest {
        headers: vec![
            ("Authorization".to_string(), authorization),
            ("x-amz-date".to_string(), amz_date),
            ("x-amz-content-sha256".to_string(), payload_hash),
        ],
    }
}
